File: SatinalmaPro.Shared/Procurement/FirestoreFilterSpec.cs
namespace SatinalmaPro.Shared.Procurement;

public sealed class FirestoreOrderBy
{
    public string Field { get; init; } = "updatedAtUtc";
    public bool Descending { get; init; } = true;
}

/// <summary>
/// Firestore sorgu şartları — platform bağımsız tanım.
/// Android: <c>FirebaseFirestore</c> · Masaüstü: REST structured query veya bellek içi filtre.
/// </summary>
public sealed class FirestoreFilterSpec
{
    public string Collection { get; init; } = "procurement_requests";
    public IReadOnlyList<string> StatusIn { get; init; } = [];
    public string? RequesterUidEquals { get; init; }
    public IReadOnlyList<FirestoreOrderBy> OrderBy { get; init; } = [new() { Field = "updatedAtUtc", Descending = true }];
    public bool UrgentFirst { get; init; }
    public bool ReadOnly { get; init; }
    public bool RequiresReturnFlag { get; init; }

    public static FirestoreFilterSpec ForStockMovements(bool readOnly = false) => new()
    {
        Collection = "stock_movements",
        OrderBy = [new() { Field = "date", Descending = true }],
        ReadOnly = readOnly
    };

    public static FirestoreFilterSpec ForStockItems(bool readOnly = true) => new()
    {
        Collection = "stock_items",
        OrderBy = [new() { Field = "materialName", Descending = false }],
        ReadOnly = readOnly
    };
}

/// <summary>Sekme eşleştirmesi için minimal talep görünümü.</summary>
public sealed class ProcurementRequestSnapshot
{
    public string Id { get; init; } = "";
    public string Status { get; init; } = "";
    public string RequesterUid { get; init; } = "";
    public string Priority { get; init; } = ProcurementPriority.Normal;
    public string RequestType { get; init; } = "Normal";

    public string NormalizedStatus => ProcurementStatus.Normalize(Status);

    public string EffectivePriority =>
        !string.IsNullOrWhiteSpace(Priority) && !Priority.Equals(ProcurementPriority.Normal, StringComparison.OrdinalIgnoreCase)
            ? Priority
            : ProcurementPriority.FromRequestType(RequestType);
}

File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/shared/filter/FirestoreFilterSpec.kt
package com.satinalmapro.shared.filter

data class FirestoreOrderBy(
    val field: String = "updatedAtUtc",
    val descending: Boolean = true
)

/**
 * Firestore sorgu şartları — platform bağımsız tanım.
 * Android: [com.satinalmapro.shared.filter.FirestoreQueryBuilder]
 */
data class FirestoreFilterSpec(
    val collection: String = "procurement_requests",
    val statusIn: List<String> = emptyList(),
    val requesterUidEquals: String? = null,
    val orderBy: List<FirestoreOrderBy> = listOf(FirestoreOrderBy()),
    val urgentFirst: Boolean = false,
    val readOnly: Boolean = false
) {
    companion object {
        fun forStockMovements(readOnly: Boolean = false) = FirestoreFilterSpec(
            collection = "stock_movements",
            orderBy = listOf(FirestoreOrderBy(field = "date", descending = true)),
            readOnly = readOnly
        )

        fun forStockItems(readOnly: Boolean = true) = FirestoreFilterSpec(
            collection = "stock_items",
            orderBy = listOf(FirestoreOrderBy(field = "materialName", descending = false)),
            readOnly = readOnly
        )
    }
}

data class ProcurementRequestSnapshot(
    val id: String = "",
    val status: String = "",
    val requesterUid: String = "",
    val priority: String = ProcurementPriority.NORMAL,
    val requestType: String = "Normal"
) {
    val normalizedStatus: String
        get() = ProcurementStatus.normalize(status)

    val effectivePriority: String
        get() = if (priority.isNotBlank() && !priority.equals(ProcurementPriority.NORMAL, ignoreCase = true)) {
            priority
        } else {
            ProcurementPriority.fromRequestType(requestType)
        }
}
